File: backend/hld_canvas_engine.py
"""Apply flowchart mutations for HLD architecture diagrams."""

from __future__ import annotations

from backend.mutation_engine import apply_mutation_mode, append_lines

HLD_CANVAS_HEADER = "flowchart TD\n"
HLD_PLACEHOLDER = 'flowchart TD\n    placeholder["Start"]\n'
PLACEHOLDER_NODE = 'placeholder["Start"]'


def init_hld_canvas() -> str:
    return HLD_PLACEHOLDER


def _apply_incremental(canvas: str, mutation: str) -> str:
    if canvas == HLD_CANVAS_HEADER or PLACEHOLDER_NODE in canvas:
        canvas = HLD_CANVAS_HEADER
    return append_lines(canvas, mutation)


def apply_hld_mutation(canvas: str, mutation: str) -> str:
    if not mutation or not mutation.strip():
        return canvas

    next_canvas, mode = apply_mutation_mode(canvas, mutation)
    if mode in ("snapshot", "style"):
        return next_canvas

    return _apply_incremental(canvas, mutation)

File: frontend/src/hldCanvasEngine.js
import { appendLines, applyMutationMode, normalizeCanvas } from "./mutationEngine.js";

const HLD_CANVAS_HEADER = "flowchart TD\n";
const HLD_PLACEHOLDER =
  'placeholder["Start"]';

export function initHldCanvas() {
  return `${HLD_CANVAS_HEADER}    ${HLD_PLACEHOLDER}\n`;
}

function applyIncremental(canvas, mutation) {
  let nextCanvas =
    canvas === HLD_CANVAS_HEADER || canvas.includes(HLD_PLACEHOLDER)
      ? HLD_CANVAS_HEADER
      : canvas;

  return appendLines(nextCanvas, mutation);
}

export function applyHldMutation(canvas, mutation) {
  if (!mutation?.trim()) return canvas;

  const { canvas: staged, mode } = applyMutationMode(canvas, mutation);
  if (mode === "snapshot") return normalizeCanvas(mutation);
  if (mode === "style") return staged;

  return applyIncremental(canvas, mutation);
}

export const HLD_METER_DEFAULTS = {
  availability: 55,
  latency: 220,
  cost: 35,
};

export const DEFAULT_HEALTH_IMPACT = {
  availability: 8,
  latency: -15,
  cost: 5,
};

export function resolveHealthImpact(apiImpact) {
  if (apiImpact && Object.keys(apiImpact).length > 0) {
    return apiImpact;
  }
  return DEFAULT_HEALTH_IMPACT;
}

export function applyHealthImpact(meters, impact) {
  if (!impact) {
    return { ...meters };
  }
  return {
    availability: clamp(
      (meters.availability ?? 0) + (impact.availability ?? 0),
      0,
      100,
    ),
    latency: clamp((meters.latency ?? 0) + (impact.latency ?? 0), 20, 500),
    cost: clamp((meters.cost ?? 0) + (impact.cost ?? 0), 0, 100),
  };
}

function clamp(value, min, max) {
  return Math.min(max, Math.max(min, value));
}

export function metersHealthy(meters) {
  return meters.availability >= 85 && meters.latency <= 80 && meters.cost <= 60;
}
